Dockerfile.crystal

```diff
@@ -22,6 +25,14 @@ RUN mkdir -p $NAV2_WS/src
 WORKDIR $NAV2_WS/src
 COPY ./ navigation2/
 
+# delete nav2_system_tests/COLCON_IGNORE before running rosdep, otherwise it
```

Comment:
Is this true? I've never known rosdep to care what a `COLCON_IGNORE` file is or what it means. I think it just recurses the `--from-paths` looking for `package.xml` files. Our CI Dockerfile is setup similarly.

Moving this arg and logic to before the `rosdep install` would needlessly break the build caching layers when switching system tests on and off.

Searching the rosdep github repo for any mentioning of `COLCON_IGNORE` returns nothing:  
https://github.com/ros-infrastructure/rosdep/search?q=COLCON_IGNORE&unscoped_q=COLCON_IGNORE

Reply:
It's true that the build breaks if the rm is done after the rosdep call. It breaks because it can't find the gazebo_ros_pkgs which are the additional dependency that the nav2_system_test package brings in.

I'll admit that I'm assuming the underlying root cause.

Reply:
I created a minimal docker to experiment with.

```
FROM ros:crystal
# install ROS2 dependencies
RUN apt-get update && apt-get install -q -y \
      build-essential \
      cmake \
      git \
      python3-colcon-common-extensions \
      python3-vcstool \
      wget \
    && rm -rf /var/lib/apt/lists/*

# copy ros package repo
ENV NAV2_WS /opt/nav2_ws
RUN mkdir -p $NAV2_WS/src
WORKDIR $NAV2_WS/src
COPY ./ navigation2/
```
With the COLCON_IGNORE file in place, I get the following output from `rosdep`

```
> rosdep install -s -q -y --from-paths src --ignore-src
#[apt] Installation commands:
  apt-get install -y -qq libboost-all-dev
  apt-get install -y -qq ros-crystal-launch-testing
  apt-get install -y -qq ros-crystal-behaviortree-cpp
  apt-get install -y -qq libbullet-dev
  apt-get install -y -qq libsdl1.2-dev
  apt-get install -y -qq libsdl-image1.2-dev
  apt-get install -y -qq ros-crystal-yaml-cpp-vendor
  apt-get install -y -qq ros-crystal-map-msgs
  apt-get install -y -qq ros-crystal-tf2-sensor-msgs
  apt-get install -y -qq libpcl-dev
  apt-get install -y -qq ros-crystal-laser-geometry
  apt-get install -y -qq ros-crystal-angles
```

If I delete the COLCON_IGNORE file

```
> rosdep install -s -q -y --from-paths src --ignore-src
#[apt] Installation commands:
  apt-get install -y -qq ros-crystal-launch-testing
  apt-get install -y -qq ros-crystal-gazebo-ros-pkgs
  apt-get install -y -qq ros-crystal-angles
  apt-get install -y -qq ros-crystal-behaviortree-cpp
  apt-get install -y -qq libboost-all-dev
  apt-get install -y -qq ros-crystal-map-msgs
  apt-get install -y -qq ros-crystal-tf2-sensor-msgs
  apt-get install -y -qq libpcl-dev
  apt-get install -y -qq ros-crystal-laser-geometry
  apt-get install -y -qq libbullet-dev
  apt-get install -y -qq libsdl1.2-dev
  apt-get install -y -qq libsdl-image1.2-dev
  apt-get install -y -qq ros-crystal-yaml-cpp-vendor
```

You'll see `ros-crystal-gazebo-ros-pkgs` is missing from the first list, so rosdep results in different action depending on whether the file is there or not. How it does that without ever actually referencing the file, I don't know.

Reply:
Ok, it looks like rosdep is using catkin's `find_packages` to loop search for packages, and thus the transitive ignore.

https://github.com/ros-infrastructure/rosdep/blob/e5d6772686874165cf1ef96d0e6b64ff63076af1/src/rosdep2/catkin_packages.py#L7